sys/linux/dev_char_raw.txt

```diff
@@ -0,0 +1,65 @@
+# Copyright 2017 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+# drivers/char/raw.c
+include <linux/raw.h>
+
+resource fd_char_raw[fd]
+resource fd_char_raw_ctl[fd]
+
+openat$char_raw_ctl(fd const[AT_FDCWD], file ptr[in, string["/dev/raw/rawctl"]], flags flags[open_flags], mode const[0]) fd_char_raw_ctl
+# /dev/raw/raw# only works if ioctl$RAW_CHAR_CTRL_SETBIND executed first.
+syz_open_dev$char_raw(file ptr[in, string["/dev/raw/raw#"]], dev intptr[1:2], flags flags[raw_open_flags]) fd_char_raw
+raw_open_flags = O_RDWR, O_RDONLY, O_APPEND, O_NOATIME, O_NOFOLLOW, O_NONBLOCK, O_PATH, O_EXCL, O_DIRECT
+
+close$char_raw(fd fd_char_raw)
+close$char_raw_ctl(fd fd_char_raw_ctl)
+
+read$char_raw(fd fd_char_raw, buf ptr[out, raw_char_buffer], count len[buf])
+write$char_raw(fd fd_char_raw, buf ptr[in, raw_char_buffer], count len[buf])
+
+raw_char_buffer {
+	buf	array[int8, 1:65536]
+} [align[512]]
+
+# IOCTL on /dev/raw/rawctl
+resource block_info_ptr[int64]
+syz_raw_bind_info(dev raw_bind_fd, raw_dev int32[1:2]) block_info_ptr
+ioctl$RAW_CHAR_CTRL_SETBIND(fd fd_char_raw_ctl, cmd const[RAW_SETBIND], arg block_info_ptr)
+ioctl$RAW_CHAR_CTRL_GETBIND(fd fd_char_raw_ctl, cmd const[RAW_GETBIND], arg block_info_ptr)
+
+resource fd_loop_file[fd]
+syz_prepare_loop_device(arg int32[1:2]) fd_loop_file
+
+raw_bind_fd [
+	fd_block	fd_block
+	fd_loop		fd_loop
+	fd_loop_file	fd_loop_file
+]
+
+# IOCTL on /dev/raw/rawN
+# equal to dev_block.txt - redefined for fd_char_raw
+include <linux/blkdev.h>
+ioctl$CHAR_RAW_FLSBUF(fd fd_char_raw, cmd const[BLKFLSBUF], arg ptr[in, intptr])
```

Comment:
Wonder if we need so much boilerplate. Filed https://github.com/google/syzkaller/issues/1913